### DISCUSSION TITLE
테스트를 왜 먼저 작성해야 할까요?

### DISCUSSION THREAD

**Discussion by hannut91** [CodeSoom/spring-week3-assignment-1#45] [Q&A]
내가 무엇을 작성해야 하는지도 모르겠는데, 테스트를 먼저 작성하라고 합니다.

왜 테스트를 먼저 작성해야 할까요?

**Comment by injoon2019** [comment]
아직 TDD에 대해 책도 읽어보지 않고, 익숙하지 않지만 귀동냥한걸 적어본다면...

1. "빠르게 실패" 함으로서 성공하기 위한 목표 지점이 반대로 명확해진다고 알고 있습니다.
2. 1번과 유사한 얘기지만 잘 쓰여진 테스트 코드는 그 자체가 하나의 문서가 될 수 있다고 알고 있습니다.

사실 아직까지 정확하게 모르겠습니다. 처음 자바를 배울 때 `객체지향이란 ~~ 것이다`라고 들었을 때 와닿지 않았지만 조금씩 알게된 것처럼 사용해보면서 알아가지 않을까 기대중입니다.

**Comment by daeungkim** [comment]
유닛테스트의 역할은 테스트라기 보다, 실제로 작동하는 주석 혹은 개발문서의 역할을 한다고 생각합니다.
해당 기능의 happy path는 어떠한지, 어떤 예외처리를 수행하는지 등을 작동하는 코드로 남겨두는것이 유닛테스트의 목적이라고 생각합니다.
이러한 유닛테스트를 먼저 작성하게 하여, 개발자가 무작정 개발에 들어가는것을 방지할 수 있습니다. 먼저 설계를 하고, 해당 설계에 기반한 개발을 강제할 수 있습니다.

**Comment by catsbi** [comment]
제가 공부한 내용은 다음과 같습니다. 

1. 테스트 자체가 하나의 API 문서가 될 수 있습니다. 
    - 작은 기능단위의 유닛테스트는 하나 하나의 테스트가 메서드를 설명하고 기댓값을 알 수 있게 해줍니다. 
    - Spring Rest Docs라는 도구를 이용하면 테스트만으로 API 문서 자동화도 가능해집니다. 
    
2. 고립 테스트를 만들기 쉬워집니다. 
    - 테스트 코드부터 작성을 하게 되면 코드에서 테스트 하고자 하는 메서드가 다른 객체와 의존성을 갖지 않는지 확인할 수 있습니다. 

3. 켄트 백의 TDD 책에선 TDD가 테스트 기술이라기보다는 분석 및 설계 기술이라고 하기도 했습니다. 

4. 한 번에 하나의 주제에만 집중할 수 있습니다. 

5. 좀 더 견고한 설계를 완성시킬 수 있습니다. 
    - Test Driven Devlopment 즉 테스트 주도 개발은 실패하는 테스트를 시작으로 구현과 개선을 반복해서 좀 더 완성된 기능을 완성할 수 있게됩니다.
